### PR TITLE
An initial ClusterProfile CRD implementation of KEP-4322: Cluster Profile API

### PR DESCRIPTION
**Pull request by mikeshng**
By community demand, this is the first go of ClusterProfile CRD implementation of [KEP 4322](https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/4322-cluster-inventory).

Note: We are expecting the PR `https://github.com/kubernetes/enhancements/pull/4533` to merge for the renaming of the API from ClusterInventory to ClusterProfile.

### CODE REVIEW

**Comment by RainbowMango** [comment]
/assign 
I'll take a look. Thank you @mikeshng .

**Comment by dixudx** [comment]
We may also need to rename this repository as well.

File: apis/v1alpha1/clusterprofile_types.go
Comment (zhiying-lin) [inline apis/v1alpha1/clusterprofile_types.go:26]: optional field?

File: apis/v1alpha1/clusterprofile_types.go
Comment (zhiying-lin) [inline apis/v1alpha1/clusterprofile_types.go:23]:
```suggestion
// ClusterProfileSpec defines the desired state of ClusterProfile.
```

File: apis/v1alpha1/clusterprofile_types.go
Comment (zhiying-lin) [inline apis/v1alpha1/clusterprofile_types.go:32]:
```suggestion
// ClusterManager defines which cluster manager created this ClusterProfile resource and it's immutable.
```

File: apis/v1alpha1/clusterprofile_types.go
Comment (zhiying-lin) [inline apis/v1alpha1/clusterprofile_types.go:48]:
```suggestion
	// Version defines the version information of the cluster.
```
we may incorporate other version, eg, enabled featureset version

File: apis/v1alpha1/clusterprofile_types.go
Comment (qiujian16) [inline apis/v1alpha1/clusterprofile_types.go:32]: s/creates/owns ?

File: apis/v1alpha1/clusterprofile_types.go
Comment (qiujian16) [inline apis/v1alpha1/clusterprofile_types.go:34]: should have a const to define this label key.

File: apis/v1alpha1/clusterprofile_types.go
Comment (qiujian16) [inline apis/v1alpha1/clusterprofile_types.go:37]: should I also indicate this fiels is immutable?

File: apis/v1alpha1/clusterprofile_types.go
Comment (qiujian16) [inline apis/v1alpha1/clusterprofile_types.go:61]:
```suggestion
	// Kubernetes is the kubernetes version of the cluster.
```

File: go.mod
Comment (RainbowMango) [inline go.mod:1]: > We may also need to rename this repository.

Good point. We need to update the module name then.

Comment (mikeshng) [reply go.mod:1]: Hi @dixudx @RainbowMango the general consensus from the SIG-MC community is that the repo name is independent from the ClusterProfile API name. The whole concept of "Cluster Inventory" is still fundamentally sound. It's just that unfortunately due to English spelling/grammar, it's not accurate to represent a single cluster using "ClusterInventory". Another reason is ClusterProfile might just be one of the APIs that represent the "Cluster Inventory" concept, so we shouldn't rename the repo to ClusterProfile but stick with the key concept which is still "Cluster Inventory".

Comment (RainbowMango) [reply go.mod:1]: Got it. I didn't know the SIG-MC already talked about the repo name. 
It makes sense that this repo is not just for `ClusterProfile`.

File: apis/v1alpha1/clusterprofile_types.go
Comment (RainbowMango) [reply apis/v1alpha1/clusterprofile_types.go:34]: +1
We should have a const for cluster manager to consume, but keep the value in comments.

File: apis/v1alpha1/clusterprofile_types.go
Comment (RainbowMango) [inline apis/v1alpha1/clusterprofile_types.go:89]:
```suggestion
//+kubebuilder:resource:scope=Cluster
```

File: hack/boilerplate.go.txt
Comment (RainbowMango) [inline hack/boilerplate.go.txt:2]: Should be?
```suggestion
Copyright 2024 The Kubernetes Authors.
```

File: config/crd/kustomization.yaml
Comment (RainbowMango) [inline config/crd/kustomization.yaml:1]: Can you remind me why we need this file?

File: PROJECT
Comment (RainbowMango) [inline PROJECT:1]: Do we need to store this file?
I checked the referencing document [Why do we need to store the plugins and data used?](https://book.kubebuilder.io/reference/project-config.html#why-do-we-need-to-store-the-plugins-and-data-used), it's unlikely for us to re-scaffolding the project in the future.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go]: Done as suggested. Thanks.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go]: Done as suggested. Thanks.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go]: Done as suggested. Thanks.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go]: Done as suggested. Thanks.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go:35]: Done as suggested. Thanks.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go:38]: Done as suggested. Thanks.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go]: Done as suggested. Thanks.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go]: Done as suggested. Thanks.

File: hack/boilerplate.go.txt
Comment (mikeshng) [reply hack/boilerplate.go.txt]: Done as suggested. Thanks.

File: config/crd/kustomization.yaml
Comment (mikeshng) [reply config/crd/kustomization.yaml:1]: Removed as suggested because it's not necessary right now. Maybe it will be useful when we have more apis in the future. Thanks.

File: PROJECT
Comment (mikeshng) [reply PROJECT:1]: Removed as suggested. Thanks.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go:26]: I think it should be required. The consumer of this API might not be able to use the ClusterProfile resource name for display purposes because if the producer follows our guideline then the resource name has be rather unique to avoid clashing. The consumer of this API will probably want to rely on having a displayName to show users the friendly name of the cluster in their UI. If it's optional and sometimes its empty then the consumer can't rely on this field as much.
WDYT?

Comment (RainbowMango) [reply apis/v1alpha1/clusterprofile_types.go:26]: +1 on required

I suggest adding `// +required` tag on each required field and adding `// +optional` on each optional field, to make it clear.

Comment (zhiying-lin) [reply apis/v1alpha1/clusterprofile_types.go:26]: based on my understanding on the field name/comment, this field is for human read only. It should be optional and depends on whether users really need this display name or not.

File: go.mod
Comment (ArangoGutierrez) [inline go.mod:3]: is there a reason not to go with latest 1.22?

**Review by ArangoGutierrez — CHANGES_REQUESTED** [review]

File: go.mod
Comment (mikeshng) [reply go.mod:3]: No particular reason. It was the version that the latest stable kubebuilder used so I didn't change it.

I will update this and bring the other dependencies to the latest stable versions as well.

File: config/crd/bases/multicluster.x-k8s.io_clusterprofiles.yaml
Comment (bsalamat) [inline config/crd/bases/multicluster.x-k8s.io_clusterprofiles.yaml:15]: is this our final decision?

Comment (mikeshng) [reply config/crd/bases/multicluster.x-k8s.io_clusterprofiles.yaml:15]: No, this is open for discussion. This PR is meant to drive some of these discussions in async manner.
We feel like Cluster scope is a more natural fit for this API so it's a good starting point.
CC @deads2k

File: apis/v1alpha1/clusterprofile_types.go
Comment (corentone) [inline apis/v1alpha1/clusterprofile_types.go:29]: more than own, its to self-identify the "creator" of the ClusterInventory object, not necessarily what/who manages it?
Does it replace the "Owner" field of the resource?

File: apis/v1alpha1/clusterprofile_types.go
Comment (corentone) [reply apis/v1alpha1/clusterprofile_types.go:40]: the label is not immutable though right? If that's the case, can we rely on the label? Or can we enforce the immutability?

File: apis/v1alpha1/clusterprofile_types.go
Comment (corentone) [reply apis/v1alpha1/clusterprofile_types.go:87]: It isn't clear what "Joined" really conveys. Doesn't the ClusterProfile CR's existence means that it has "Joined" the list of ClusterProfiles?

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go:29]: It doesn't replace the Owner field. This field is to indicate which cluster manager this resource came from so that if it's an environment with multiple cluster managers, then each manager knows which resources to manage.
I had created in the beginning but it was suggested to change to own: https://github.com/kubernetes-sigs/cluster-inventory-api/pull/6#discussion_r1568165541
WDYT? How should I best describe this?

File: go.mod
Comment (mikeshng) [reply go.mod]: Done as suggested. Thanks.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go]: Good catch. Done as suggested. Thanks.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go]: Yes, this is a confusing condition. Originally, we wanted a condition to highlight the fact that a cluster is now "registered" into a cluster manager. I guess we are not ready to flush out all that details yet so I removed this constant for now.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go:99]: I've removed the Joined condition for now to avoid the confusion. See https://github.com/kubernetes-sigs/cluster-inventory-api/pull/6#discussion_r1572638760 for more details.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go:82]: This was heavily debated during a few SIG MC meetings. If we include real time data such as CPU, GPU, memory etc. it might not scale well when the cluster manager watches many clusters.
For now, I've adjusted the comment to mention that it may include these data because I think it's still up the user of this API. WDYT?

File: apis/v1alpha1/clusterprofile_types.go
Comment (bsalamat) [reply apis/v1alpha1/clusterprofile_types.go:99]: This is less confusing. We may want to add other conditions for "unhealthy" cases, for example, we may want to have a "disconnected" status for a cluster that is registered to a fleet, but it cannot be reached.

File: apis/v1alpha1/clusterprofile_types.go
Comment (corentone) [reply apis/v1alpha1/clusterprofile_types.go:29]: Thank you for the details.

I see you added a bit more details on ClusterManager. I think I understand better your intent.
Now, it sounds like this field is only a hint for the managers to not step on each other (I don't think it will be permission protected right?). If we assume it's only a "hint", why keep it immutable and required?

Could a cluster manager transfer management to another one (or an admin deciding it wants to transition it?), Could we have a situation where the clsuter profile is manually created and therefore has no manager?
If we push it in this direction (optional and mutable), why not just using the label?

Does knowing what the manager is convey information to consumers of the ClusterProfile object?

Comment (bsalamat) [reply apis/v1alpha1/clusterprofile_types.go:29]: I am not worried about changing the cluster manager. One could argue that the ClusterProfile should be deleted and recreated in order to change the manager. In that case ClusterManager can remain immutable. What worries me more is that we allow only one ClusterManager for each member cluster. Having only one manager simplifies certain operations, but I wonder if it is too limiting. What if various operations of a fleet are performed by multiple ClusterManagers?

Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go:29]: > What if various operations of a fleet are performed by multiple ClusterManagers?

ClusterProfile is meant to be continuously updated by one ClusterManager. If we allow multiple ClusterManagers to update it then I think it will be hard to coordinate who is allow to update which fields without stepping on each other like Corentin said.

> One could argue that the ClusterProfile should be deleted and recreated in order to change the manager. In that case ClusterManager can remain immutable.

This is a great point. If we can come to a conclusion on this I will add it to the field description.

Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go:29]: > I don't think it will be permission protected right?

We do have the CEL validation for now to block it from being updated.

> If we assume it's only a "hint", why keep it immutable and required?

I think it's important to prevent multiple ClusterManagers from updating the same ClusterProfile so consumers can get a consistent feed of data from one source. 

> Could a cluster manager transfer management to another one

I like Babak comment above regarding delete and recreate but I hope others can chime in on this as well. CC @qiujian16 @dixudx 

> Does knowing what the manager is convey information to consumers of the ClusterProfile object?

This is a good point. As you suggested, this information seems to be only useful for the producers but I am thinking it might be useful in some cases for the consumer as well. For example, maybe the consumer can be the provider as well. Maybe ArgoCD can have it's own cluster inventory which produces ClusterProfile objects but also have the mechanism to import ClusterProfiles into its own registry. In that case, it might be helpful to know where the ClusterProfile it came from? See https://github.com/argoproj-labs/argocd-operator/issues/957

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go:34]: Done as suggested. Expanded what a Cluster Manager means. Thanks.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go:99]: How about "available" and "unavailable" instead of healthy/unhealthy/connected/disconnected? 

In the SIG MC meetings and the KEP PR, we had multiple lengthy discussions what healthy and unhealthy means. I feel like available and unavailable is more generic for the consumers instead of diving deep into what healthy and connection status means for now.

Comment (bsalamat) [reply apis/v1alpha1/clusterprofile_types.go:99]: 'available/unavailable' sounds good to me.

File: apis/v1alpha1/clusterprofile_types.go
Comment (bsalamat) [reply apis/v1alpha1/clusterprofile_types.go:29]: Assuming that changing the cluster manager is a rare event, keeping it immutable is a good idea. That way the cluster managers know that the list of their managed clusters won't change and they only need to watch delete/create events to update the list.

File: config/crd/bases/multicluster.x-k8s.io_clusterprofiles.yaml
Comment (bsalamat) [reply config/crd/bases/multicluster.x-k8s.io_clusterprofiles.yaml:15]: My general guidance for creating new CRDs is that "if the CRD doesn't have to be cluster-scoped, it should be namespace-scoped."
Also, if we had a namespace with a predefined name for all multi-cluster management objects, such as cluster profile, and other future objects, creating RBAC rules for accessing them in that namespace would be simpler.

File: apis/v1alpha1/clusterprofile_types.go
Comment (corentone) [reply apis/v1alpha1/clusterprofile_types.go:29]: Thank you for the long answer.
CEL validation for the label is good.

Explanation and extra details for clusterManager helped. I think they would "own" the object since it's immutable, but the result is the same.
Immutable is good and possible to relax (while going the other way is not possible).

For specific parameters where we fear overlap, we can have additional rules (kinda similar to scheduler owning a subset of fields).

Change looks good for me :)

File: config/crd/bases/multicluster.x-k8s.io_clusterprofiles.yaml
Comment (thockin) [reply config/crd/bases/multicluster.x-k8s.io_clusterprofiles.yaml:15]: I don't have a _strong_ feeling, mostly because we just don't know what we don't know, and need to explore it.

That said, my instinct is that if you CAN namespace it, you probably should, unless there's a good reason not to.  We have plenty of examples of cluster-scoped things we wish we had namespaced, and few to none of the reverse examples.

Comment (corentone) [reply config/crd/bases/multicluster.x-k8s.io_clusterprofiles.yaml:15]: Discussion is happening here: https://docs.google.com/document/d/1E_z3ti-d-modwnhsvR3yBZwX4rRpL26dKkl41YAptRo/edit

**Review by ArangoGutierrez — CHANGES_REQUESTED** [review]
could you squash the commits into one once we are done. 4 commits, with 2 being "address feedback" is not something we want for the git history of a new project

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikemorris) [reply apis/v1alpha1/clusterprofile_types.go]: No strong feelings on if this is actually needed, but if we do want to include a status like this `Registered` (or `Accepted` which is commonly used for resources over in Gateway API) feel like better options than `Joined`.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikemorris) [reply apis/v1alpha1/clusterprofile_types.go:29]: > If we allow multiple ClusterManagers to update it then I think it will be hard to coordinate who is allow to update which fields without stepping on each other like Corentin said.

If there's a need to support multiple controllers updating one ClusterProfile resource, we could look to the pattern Gateway API uses for [RouteStatus](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.RouteParentStatus), with a separate set of status conditions scoped to each controller. I'm not sure if this capability is actually needed though, or if it would be better to keep the ClusterManager relationship simpler.

**Review by mikemorris — COMMENTED** [review]
This looks like a great start! I think there's a few questions about intended supportable functionality or use cases that would help clarify some of the open spec design questions.

- [ ] Should it be possible to host multiple inventories on a single hub cluster?
- [ ] Is there a valid use case for multiple cluster managers wanting to update a single ClusterProfile?
- [ ] Would a ClusterProfile would ever be created directly by an end user instead of automatically through a cluster manager?
- [ ] Is a ClusterProfile only intended to be present on a "hub" cluster, or is there a pattern any implementer wants that would have the ClusterProfile present on a member cluster?

File: apis/v1alpha1/clusterprofile_types.go
Comment (corentone) [reply apis/v1alpha1/clusterprofile_types.go:44]: We can close this based on other reply. CEL validation will enforce it.

File: apis/v1alpha1/clusterprofile_types.go
Comment (thockin) [reply apis/v1alpha1/clusterprofile_types.go:126]: This looks like the usual pattern for defining a k8s type.  E.g https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L5472

Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go:126]: Resolving this one. Feel free to reopen if needed.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go:87]: Good point Mike. I've reworked the description to mention that it could be a ClusterProperty (or something else). PTAL thanks.

File: apis/v1alpha1/clusterprofile_types.go
Comment (ryanzhang-oss) [reply apis/v1alpha1/clusterprofile_types.go:29]: I don't think we restrict one ClusterManager to have only one controller (i.e. we can have one controller collect cpu/memory property while another one collect ARD/Device property in the same clusterManager). However, I don't know if the benefit of supporting multiple ClusterManager for a single ClusterProfile worth breaking the neat immutability of the clusterManager field.

File: apis/v1alpha1/clusterprofile_types.go
Comment (ryanzhang-oss) [reply apis/v1alpha1/clusterprofile_types.go:87]: The current thinking is to keep the property field open while the community comes to an agreement on what are the "core" properties every provider should support.  There are also "nice to have", "you are on your own" properties that we can list in the KEP/CRD.

Comment (mikemorris) [reply apis/v1alpha1/clusterprofile_types.go:87]: Revised description looks good enough to get started, okay to resolve this thread I think!

**Review by ArangoGutierrez — CHANGES_REQUESTED** [review]

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go:30]: We had multiple lengthy SIG-MC meeting discussions on how to avoid Cluster Managers from stepping on each other and the current solution is to make sure we know which Cluster Manager owns a ClusterProfile. WDYT?

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go]: I've removed this condition for now . If we want to bring it back as Registered or Accepted (prefer accepted because it's used in gateway API), we can reopen this thread.

File: apis/v1alpha1/clusterprofile_types.go
Comment (bsalamat) [reply apis/v1alpha1/clusterprofile_types.go]: After thinking more about this, I am still not clear what it signifies. I assume, there will be a controller somewhere that updates this status. The controller may be running on the manager or on the member clusters. When we set the status to "available" does it mean that the control plane of the cluster is available? Or, does it mean that the data plane of the cluster is "available"?
Practical scenarios:
1. The controller which sets this status runs on the control plane of the cluster and sends us a heartbeat. It makes the cluster look available while its data plane might be down.
2. The controller runs on a node in the member cluster and sends us a heartbeat. The cluster will look available even when its control plane is down.
3. The controller runs on the manager and checks a "ping" endpoint, depending on where that endpoint runs, we will have one of the above scenarios.

An alternative to the above is to have more logic that checks both control plane and data plane and marks the cluster available only when both are functional.
This long comment is here to just say that we should define these conditions more precisely.

File: apis/v1alpha1/clusterprofile_types.go
Comment (bsalamat) [inline apis/v1alpha1/clusterprofile_types.go:60]: This field is required, but ClusterVersion has only one optional field. Should this be optional, or should the only field of ClusterVersion be required?

Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go]: Good catch. This should be optional. There might be a scenario that we know a cluster is in the inventory but we don't know the specific version or other version info yet. Fixed.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go]: Reworked this to be `ControlPlaneHealthy`. PTAL. 

The data plane healthiness condition is a lot more difficult to define. It will probably require it's own separate PR. CC @ArangoGutierrez

Comment (bsalamat) [reply apis/v1alpha1/clusterprofile_types.go]: Thank you. It is clear now. "Conditions" is a required field of ClusterProfileStatus and we currently have only this predefined condition. Do we expect this condition to be always present in ClusterProfileStatus?

Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go]: Yes, I think it's important for the consumers to be able to rely on the existence of the conditions field. If the cluster manager is unable to determine the `ControlPlaneHealthy`, setting the status to `Unknown` is acceptable.

File: apis/v1alpha1/clusterprofile_types.go
Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go:29]: Resolving this for now. We defined what a cluster manager is and decided that it should be an immutable field for now.

File: apis/v1alpha1/clusterprofile_types.go
Comment (ArangoGutierrez) [reply apis/v1alpha1/clusterprofile_types.go:30]: I guess the controller would have to set it by default to none or to the provider of the controller. but ok, I agree on having it `required`

File: apis/v1alpha1/clusterprofile_types.go
Comment (thockin) [reply apis/v1alpha1/clusterprofile_types.go]: Spec and status are updated independently.  Having `Conditions` be required means there MUST BE a condition specified when the ClusterProfile is created.  This is not something I have ever seen before, and I would argue that if you have a required condition, you should just use a proper field.

Comment (mikeshng) [reply apis/v1alpha1/clusterprofile_types.go]: You are right. Based on the core types https://github.com/kubernetes/api/blob/master/core/v1/types.go all the condition fields are optional. I am going to fix this. 
My original thinking was Status is marked optional so if a status is populated then a required condition inside the status seems reasonable.

**Comment by mikeshng** [comment]
> This looks like a great start! I think there's a few questions about intended supportable functionality or use cases that would help clarify some of the open spec design questions.

We are making some KEP changes to clarify these design questions. This implementation PR will follow the KEP once it's merged.

File: config/crd/bases/multicluster.x-k8s.io_clusterprofiles.yaml
Comment (mikeshng) [reply config/crd/bases/multicluster.x-k8s.io_clusterprofiles.yaml:15]: KEP PR to clarify the scope of ClusterProfile: https://github.com/kubernetes/enhancements/pull/4620
This CRD implementation will follow the KEP once it's merged.

Comment (ArangoGutierrez) [reply config/crd/bases/multicluster.x-k8s.io_clusterprofiles.yaml:15]: thanks for the update @mikeshng

Comment (mikeshng) [reply config/crd/bases/multicluster.x-k8s.io_clusterprofiles.yaml]: Updated to namespaced scope. PTAL.

**Review by ryanzhang-oss — APPROVED** [review]

**Pull request merged**
